Embedded artwork support

Saecki opened the pull request:
This enables embedded artwork support (#34) for `id3` and `mp4` tags.

agersant commented on src/artwork.rs:
```diff
@@ -0,0 +1,74 @@
+fn read_mp4_artwork(path: &Path) -> Result<DynamicImage> {
+	let tag = mp4ameta::Tag::read_from_path(path)?;
+
+	match tag.artwork() {
```
Can you emit a different error for `Some(_)` vs `None` in case there are other image formats to support in the future?

agersant commented on src/index/metadata.rs:
```diff
@@ -24,6 +24,7 @@ pub struct SongTags {
 	pub album_artist: Option<String>,
 	pub album: Option<String>,
 	pub year: Option<i32>,
+	pub artwork: bool,
```
Nitpick: can you rename this field to `has_artwork`?

agersant commented on src/index/update.rs:
```diff
@@ -223,6 +223,12 @@ impl IndexUpdater {
 				directory_artist = tags.artist.as_ref().cloned();
 			}
 
+			let artwork_path = if tags.artwork {
```
I see an issue where a collection using only embedded album art would end up with no artwork associated with directories at all. That is because the `// Insert directory` block below ignores the artwork found in tags.

I would suggest using the first embedded artwork found in a song for the directory artwork.

agersant commented:
This looks great, thanks a ton for doing this work.

In addition to the inline comments above, I would prefer to have unit tests added before merging. Could you add:
- A test in `metadata.rs` validating that the `has_artwork` field of `SongTags` is populated correctly for ID3, FLAC and mp4 files with embedded art. You can add example files for these tests under a new `test-data/artwork` directory.
- A test in `index/test.rs` validating that indexed songs and directories have their artwork field set to the correct file.
- A test in `artwork.rs` validating that `read_artwork` produces images for supported formats (just validating image existence and dimensions is fine, no need to do content comparisons).

Saecki commented:
Alright I added some sample files in `test-data/artwork`, added a test in `artwork.rs` and modified the test in `index/metadata.rs`. And I embedded the `Folder.jpg` image from `Tobokegao/Picnic/`into `07 - なぜ (Why).mp3` and added a test to `index/test.rs` to validate the artwork path.

Saecki commented:
Are there reasons why `Cargo.lock` isn't included in the `.gitignore`? If not could it be added to clean up commits?

agersant commented:
> Are there reasons why `Cargo.lock` isn't included in the `.gitignore`? If not could it be added to clean up commits?

That is intentional. Reasoning behind it is [here](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries.)

agersant commented on src/index/test.rs:
```diff
@@ -56,6 +56,37 @@ fn test_metadata() {
+
+	assert_eq!(songs.len(), 1);
+	let song = &songs[0];
+	assert_eq!(song.path, song_path.to_string_lossy().as_ref());
```
No need to validate all the fields other than artwork here, the test above covers that.

agersant commented on src/index/update.rs:
```diff
@@ -226,9 +226,13 @@ impl IndexUpdater {
 			};
 
+			if directory_artwork.is_none() {
+				directory_artwork = artwork_path.as_ref().cloned();
```
This does solve the previous issue where directories would end up with no artwork at all, but it creates a new problem where artwork from a song may propagate partially to songs in the directory. Consider a directory with no Folder image, and the following content:

- Song 1 (no embedded art)
- Song 2 (embedded art)
- Song 3  (no embedded art)

Assuming the songs get processed in this order, we would end up with art for songs 2 and 3 but not for song 1. I think we should add a preliminary loop when `directory_artwork` is `None` in order to pick a `directory_artwork` that can be applied consistently to all songs (unless they have their own embedded artwork).

agersant commented:
Ship it!
